x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/rate/RateAggregationBuilder.java

```diff
@@ -195,6 +198,22 @@ static Rounding.DateTimeUnit parse(String rateUnit) {
 
     @Override
     protected ValuesSourceConfig resolveConfig(AggregationContext context) {
+        if (field() != null) {
+            MappedFieldType ft = context.getFieldType(field());
+            if (ft.getMetricType() == TimeSeriesParams.MetricType.counter) {
+                return new ValuesSourceConfig(
+                    TimeSeriesValuesSourceType.COUNTER,
```

Comment:
I don't love this.  I'd rather that the `IndexFieldData` object that this field eventually generates would report its `ValuesSourceType` as `TimeSeriesValuesSourceType.COUNTER`.  That does mean that we'd have to explicitly add `COUNTER` as a supported values source type for other aggregations, but that's not too hard.

My thinking is that we will eventually want to specialize other aggregations for time series, and I don't want to have a custom `resolveConfig` each time we want to do that.  Indeed, that kind of goes against the entire point of the `ValuesSourceRegistry`.

Having said that, I'm not sure how much work it would be to move this information into `IndexFieldData`.  Happy to discuss further if there are concerns around that.

Reply:
An additional wrinkle is that we ideally want to only select this agg implementation if `rate` is nested under a `tsid` agg, and @iverase and I thought this would be the best place to do that - extend `resolveConfig` to take the parent agg builder and then iterate through parents to see if there's a `tsid` agg somewhere in the tree.